Short-circuit methods for lambdas from annotation

artembilan opened the pull request:
When we have a `@ServiceActivator` or any other messaging annotations
on `Function` or `Consumer` `@Bean`s, there is a restriction when we
can't use lambdas because of target method argument type erasure in Java.

* Use a `@Bean` method return to determine the target function argument
type and wrap the call into the `LambdaMessageProcessor`.

In this case we call the target method directly after possible payload
conversion according expected generic type for `Function` or `Consumer`.
There is just no reason to go a `MessagingMethodInvokerHelper` route
for this lambda variants

<!--
Thanks for contributing to Spring Integration. Please provide a brief description of your pull-request and reference any related issue numbers (prefix references with #).

See the [Contributor Guidelines for more information](https://github.com/spring-projects/spring-integration/blob/master/CONTRIBUTING.adoc).
-->

artembilan commented:
If the solution is OK, I'll go ahead with other Messaging annotations.
Also need to check with Kotlin Lambdas and, of course, really check if the target classs is lambda...

garyrussell commented on spring-integration-core/src/main/java/org/springframework/integration/config/annotation/ServiceActivatorAnnotationPostProcessor.java:
```diff
@@ -63,7 +67,17 @@ protected MessageHandler createHandler(Object bean, Method method, List<Annotati
+
+						Class<?> expectedPayloadType = methodReturnType.getGeneric(0).toClass();
+						serviceActivator = new ServiceActivatingHandler(
+								new LambdaMessageProcessor(target, expectedPayloadType));
```
I think you need to check for wildcard generic types here.

artembilan replied:
Not sure what you mean, but with the `methodReturnType.getGeneric(0).toClass()` it returns an `Object` class for a `Consumer<?>`. So, I think we are good.

garyrussell reviewed (dismissed):
Nice! Just one comment.

artembilan committed:
Short-circuit methods for lambdas from annotation

When we have a `@ServiceActivator` or any other messaging annotations
on `Function` or `Consumer` `@Bean`s, there is a restriction when we
can't use lambdas because of target method argument type erasure in Java.

* Use a `@Bean` method return to determine the target function argument
type and wrap the call into the `LambdaMessageProcessor`.

In this case we call the target method directly after possible payload
conversion according expected generic type for `Function` or `Consumer`.
There is just no reason to go a `MessagingMethodInvokerHelper` route
for this lambda variants

* Apply the short-circuit algorithm for Kotlin lambdas as well
* Make some refactoring and improvements to `ClassUtils`
if favor or similar API in the SF `ClassUtils`
* Fix `No beanFactory` warning for the `ExpressionCommandMessageProcessor`

artembilan commented:
Ready for review.

Thanks

garyrussell commented:
```
> Task :spring-integration-core:checkstyleMain
[ant:checkstyle] [ERROR] /home/travis/build/spring-projects/spring-integration/spring-integration-core/src/main/java/org/springframework/integration/config/annotation/ServiceActivatorAnnotationPostProcessor.java:23:8: Unused import - java.util.function.Consumer. [UnusedImports]
[ant:checkstyle] [ERROR] /home/travis/build/spring-projects/spring-integration/spring-integration-core/src/main/java/org/springframework/integration/config/annotation/ServiceActivatorAnnotationPostProcessor.java:24:8: Unused import - java.util.function.Function. [UnusedImports]
[ant:checkstyle] [ERROR] /home/travis/build/spring-projects/spring-integration/spring-integration-core/src/main/java/org/springframework/integration/config/annotation/ServiceActivatorAnnotationPostProcessor.java:28:8: Unused import - org.springframework.core.ResolvableType. [UnusedImports]
[ant:checkstyle] [ERROR] /home/travis/build/spring-projects/spring-integration/spring-integration-core/src/main/java/org/springframework/integration/config/annotation/ServiceActivatorAnnotationPostProcessor.java:32:8: Unused import - org.springframework.integration.handler.LambdaMessageProcessor. [UnusedImports]
[ant:checkstyle] [ERROR] /home/travis/build/spring-projects/spring-integration/spring-integration-core/src/main/java/org/springframework/integration/handler/ExpressionCommandMessageProcessor.java:55:17: Reference to instance variable 'methodFilter' needs "this.". [RequireThis]
[ant:checkstyle] [ERROR] /home/travis/build/spring-projects/spring-integration/spring-integration-core/src/main/java/org/springframework/integration/util/ClassUtils.java:20:8: Unused import - java.util.HashMap. [UnusedImports]
[ant:checkstyle] [ERROR] /home/travis/build/spring-projects/spring-integration/spring-integration-core/src/main/java/org/springframework/integration/util/ClassUtils.java:21:8: Unused import - java.util.Map. [UnusedImports]
> Task :spring-integration-core:checkstyleMain FAILED
```

artembilan committed:
* Resolve Checkstyle violations

artembilan commented:
Pushed the fix.

Thank you for pointing that out!

There is an `AnnotationBeanUtils` deprecated in SF, but that should be considered as a separate issue...

garyrussell commented on spring-integration-core/src/main/java/org/springframework/integration/config/annotation/AbstractMethodAnnotationPostProcessor.java:
```diff
@@ -510,9 +515,26 @@ protected void checkMessageHandlerAttributes(String handlerBeanName, List<Annota
 	}
 
+	protected boolean resolveAttributeToBoolean(String requiresReply) {
+		return Boolean.parseBoolean(this.beanFactory.resolveEmbeddedValue(requiresReply));
```
Sounds like a generic method name, but specific parameter name. `String attribute` ?

garyrussell commented on spring-integration-core/src/main/java/org/springframework/integration/dsl/IntegrationFlowDefinition.java:
```diff
@@ -3118,7 +3122,8 @@ protected StandardIntegrationFlow get() {
 				throw new BeanCreationException("The 'currentMessageChannel' (" + this.currentMessageChannel
-						+ ") is a prototype for 'FixedSubscriberChannel' which can't be created without 'MessageHandler' "
+						+ ") is a prototype for 'FixedSubscriberChannel' which can't be created without " +
+						"'MessageHandler' "
```
without **a** 'MessageHandler'.

garyrussell reviewed (changes requested):
Just a couple of minor issues.

artembilan committed:
* Fix `resolveAttributeToBoolean()` argument name to be generic
* Fix typo in the exception message for `IntegrationFlowDefinition.get()`

garyrussell commented:
```
org.springframework.integration.jmx.config.OperationInvokingOutboundGatewayTests > gatewayWithPrimitiveArgs FAILED
    org.springframework.messaging.MessagingException at OperationInvokingOutboundGatewayTests.java:103
```

artembilan committed:
* Revert `ClassUtils.resolvePrimitiveType()` logic: an existing in SF
does exactly opposite one
